File: LeetCode/88. 合并两个有序数组.py
# 双指针
class Solution:
    def merge(self, nums1: list[int], m: int, nums2: list[int], n: int) -> None:
        nums = list()
        p1 = p2 = 0
        while p1 < m and p2 < n:
            if nums1[p1] > nums2[p2]:
                nums.append(nums2[p2])
                p2 += 1
            else:
                nums.append(nums1[p1])
                p1 += 1

        if p1 == m:
            nums += nums2[p2:n]
        else:
            nums += nums1[p1:m]

        nums1[:] = nums

# # 合并排序
# class Solution:
#     def merge(self, nums1: list[int], m: int, nums2: list[int], n: int) -> None:
#         nums1[m:] = nums2
#         nums1.sort()

File: LeetCode/88-合并两个有序数组.cpp
class Solution
{
public:
    void merge(vector<int> &nums1, int m, vector<int> &nums2, int n)
    {
        vector<int> result;
        int i = 0, j = 0;
        while (i < m || j < n)
        {
            // i < m: 在i >= m && j < n的情况下会判断nums1[i] < nums2[j]，但报错溢出
            if (j >= n || (i < m && nums1[i] < nums2[j]))
            {
                result.push_back(nums1[i]);
                ++i;
            }
            else
            {
                result.push_back(nums2[j]);
                ++j;
            }
        }
        for (int k = 0; k < nums1.size(); ++k)
        {
            nums1[k] = result[k];
        }
    }
};
